<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Derek Pappas, Ph.D. (dpappas@chori.org)" />

<meta name="date" content="2020-10-07" />

<title>BIGDAWG</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>
<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
      </style>





<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">BIGDAWG</h1>
<h4 class="author">Derek Pappas, Ph.D. (<a href="mailto:dpappas@chori.org" class="email">dpappas@chori.org</a>)</h4>
<h4 class="date">2020-10-07</h4>



<ul>
<li>Package Version: 2.3.1</li>
<li>HLA/IMGT Database Release 3.38.0, 2019-10-17</li>
</ul>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>‘Bridging ImmunoGenomic Data-Analysis Workflow Gaps’ (‘BIGDAWG’) is an integrated analysis system that automates the manual data-manipulation and trafficking steps (the gaps in an analysis workflow) normally required for analyses of highly polymorphic genetic systems (e.g., the immunological human leukocyte antigen (HLA) and killer-cell Immunoglobulin-like receptor (KIR) genes) and their respective genomic data (immunogenomic) (Pappas DJ, Marin W, Hollenbach JA, Mack SJ. 2016. ‘Bridging ImmunoGenomic Data Analysis Workflow Gaps (BIGDAWG): An integrated case-control analysis pipeline.’ <a href="https://pubmed.ncbi.nlm.nih.gov/26708359/">Human Immunology. 77:283-287</a>). Starting with unambiguous genotype data for case-control groups, ‘BIGDAWG’ performs tests of Hardy-Weinberg equilibrium, and carries out case-control association analyses for haplotypes, individual loci, and HLA amino acid positions.</p>
</div>
<div id="input-data" class="section level2">
<h2>Input Data</h2>
<p>Data for BIGDAWG should be in a tab delimited text format (UTF-8 encoding). The first row must be a header line and must include column names for genotype data. The first two columns must contain subject IDs and phenotypes (0 = control, 1 = case), respectively. A phenotype is not limited to disease status and may include other phenotypes such as onset, severity, ancestry, etc. However, phenotype designatons in the dataset are restricted to the use of 0s and 1s. Genotype pairs must be located in adjacent columns. Column names for a given locus may use ’_1’, ‘.1’,’_2’,‘.2’ to distinguish each locus pair. Genotype calls may include any text (numeric or character) except the numbers 1 and 0.</p>
<p>Data may also be passed to BIGDAWG as an R object (dataframe) following the same formatting as above for text files. You may also choose to run a synthetic HLA data set (see below) to observe a typical BIGDAWG analysis and experiment with parameter settings.</p>
<p>For HLA alleles, you may choose to format your genotype calls with our without the locus prefix. For example, for HLA-A, a given genotype call maybe 01:01:01:01 or A*01:01:01:01 or HLA-A*01:01:01:01. Allele names can include any level of resolution, from a single field up to the full length name. For HLA-DRB3,-DRB4,-DRB5 genotype calls, you may choose to represent these as a single pair of columns or as separate pairs of columns for each locus. However, when submitted as a single pair of columns, all genotypes must be formatted as Locus*Allele (including non-DRB loci). The single pair column names may be DRB345, DRB3.4.5 or DRB3/4/5. Homozygous or hemizygous status for DRB3, DRB4 and DRB5 genotypes is based on the DRB1 haplotype as defined by Andersson, 1998 (Andersson G. 1998. Evolution of the HLA-DR region. <a href="https://pubmed.ncbi.nlm.nih.gov/9675159/">Front Biosci. 3:d739-45</a>. If you wish to define your own zygosity, it is suggested you split them into separate pairs of columns for each locus manually.</p>
<p><strong>Missing Information</strong><br />
When there is missing information, either for lack of genotyping information or absence of genotyped loci, BIGDAWG allows for conventions to differentiate the type of data that is missing.</p>
<p>Data missing due to lack of a molecular genotyping result is considered not available (NA). Acceptable NA strings include: NA, ****, -, na and Na. Empty data cells will be considered NA.</p>
<p>Data missing due to genomic structural variation (i.e., no locus present) is considered absence. Acceptable absence strings include: Absent, absent, Abs, ABS, ab, Ab, AB, @. The last symbol is the unicode at sign. BIGDAWG allows for a special allele name that indicates absence of an HLA locus: 00, 00:00, 00:00:00 and 00:00:00:00 are all acceptable indicators of HLA locus absence. When choosing to use 00’s (zeros) to populate allele name fields, use similar or higher levels of resolution <a href="http://hla.alleles.org/nomenclature/naming.html">http://hla.alleles.org</a> and follow the same naming convention as with other genotype calls (either with or without locus prefix). If using a single column pair for DRB3/4/5 and the “00” absence indicator, then do <strong>NOT</strong> affix a locus prefix for the absent calls. In this case, only include the locus prefix for known DRB345 genotypes (i.e., DRB3/4/5*00:00 is <strong>NOT</strong> an acceptable name). For HLA data, the 00:00 naming convention is preferred and all absent designations will be converted to allow the amino acid analysis to test phenotypic associations with locus absence (see below).</p>
<p>Finally, when missing alleles (due to lack of a genotype call and not locus absence) for a locus are included in the haplotype analysis, the haplotype estimation method may impute the identity of the missing alleles for that subject. If such imputation is not desired, the “Missing”&quot; parameter should be set to 0.</p>
<p><strong>Genotype List Strings</strong><br />
For for HLA alleles, you may submit your data formatted as a genotype list strings (GL strings) and BIGDAWG can automatically convert the data to a tabular format. The data should be 3-columns with the GL string in the third column (see table below). Data is restricted to <strong>00:00</strong> for absent designations when using GL strings, you should NOT use any other indicator of absence. Utilizing NA is not compatible with GL2Tab conversion.</p>
<p>BIGDAWG also has a built-in function for converting between GL strings and tabular formatting using the GLSconvert() function. Please see the GLSconvert vignette for more detail.</p>
<p><strong>Novel Alleles</strong><br />
BIGDAWG will accept novel allele names. However, it is suggested you follow the same naming convention for novel alleles as with other genotypes calls in your data, either with or without the locus prefix. For example, novel alleles could be submitted as follows: Novel, 01:Novel, or A*01:Novel. Unfortunately, the BIGDAWG amino acid analysis cannot accept novel allele designations and will display an error. If you would like to run the amino acid analysis, you should replace the novel allele with NA or omit the subject entirely.</p>
<p><strong>Example of data architecture and acceptable values:</strong></p>
<p><em>Tabular</em></p>
<table>
<thead>
<tr class="header">
<th>SubjectID</th>
<th align="center">Disease</th>
<th align="center">A</th>
<th align="center">A</th>
<th align="center">B</th>
<th align="center">B</th>
<th align="center">DRB1</th>
<th align="center">DRB1</th>
<th align="center">DRB3</th>
<th align="center">DRB3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>subject1</td>
<td align="center">0</td>
<td align="center">01:01</td>
<td align="center">02:01</td>
<td align="center">08:01</td>
<td align="center">44:02</td>
<td align="center">01:01</td>
<td align="center">08:01</td>
<td align="center">Abs</td>
<td align="center">Abs</td>
</tr>
<tr class="even">
<td>subject2</td>
<td align="center">1</td>
<td align="center">02:01</td>
<td align="center">24:02</td>
<td align="center">51:01</td>
<td align="center">51:01</td>
<td align="center">11:01</td>
<td align="center">14:01</td>
<td align="center">02:02</td>
<td align="center">02:11</td>
</tr>
<tr class="odd">
<td>subject3</td>
<td align="center">0</td>
<td align="center">03:01</td>
<td align="center">26:02</td>
<td align="center">NA</td>
<td align="center">NA</td>
<td align="center">10:01</td>
<td align="center">08:01</td>
<td align="center">00:00</td>
<td align="center">00:00</td>
</tr>
</tbody>
</table>
<p><em>Genotype List String</em></p>
<table>
<colgroup>
<col width="12%"></col>
<col width="9%"></col>
<col width="77%"></col>
</colgroup>
<thead>
<tr class="header">
<th>SubjectID</th>
<th align="center">Disease</th>
<th align="center">GLString</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Subject1</td>
<td align="center">0</td>
<td align="center">HLA-A*01:01+HLA-A*02:01^HLA-B*08:01+HLA-B*44:02^HLA-DRB1*01:01+HLA-DRB1*03:01^HLA-DRB3*00:00+HLA-DRB3*00:00</td>
</tr>
<tr class="even">
<td>subject2</td>
<td align="center">1</td>
<td align="center">HLA-A*02:01+HLA-A*24:02^HLA-B*51:01+HLA-B*51:01^HLA-DRB1*11:01+HLA-DRB1*14:01^HLA-DRB3*02:02+HLA-DRB3*02:11</td>
</tr>
<tr class="odd">
<td>subject3</td>
<td align="center">0</td>
<td align="center">HLA-A*01:01+HLA-A*02:01^HLA-DRB1*01:01+HLA-DRB1*03:01^HLA-DRB3*00:00+HLA-DRB3*00:00</td>
</tr>
</tbody>
</table>
</div>
<div id="data-output" class="section level2">
<h2>Data Output</h2>
<p>After the package is run, BIGDAWG will create a new folder entitled ‘output hhmmss ddmmyy’ in the working directory (unless otherwise specified by Results.Dir parameter, see below). Within the output folder will be a precheck file (‘Data_Summary.txt’) detailing the summary statistics of the dataset and the results of the Hardy-Weinberg equilibrium test (‘HWE.txt’). If any errors are present, a log file (‘Error_Log.txt’) will be written. If no locus subsets are specified (see parameters section), a single subfolder entitled ‘Set1’ will contain the outputs of each association analysis run. If multiple locus subsets are defined, subfolders for each locus set will be created containing the respective analytic results for that subset. Within each set subfolder, a parameter file will detail the parameters that are relevant to that subset.</p>
<p>When all pairwise combinations are run in the haplotype analysis, each pairwise set will be written to a single file. A separate file called “haplotype_PairwiseSets.txt” will be written to the corresponding set’s directory and will breakdown the locus make up for each PairwiseSet.</p>
<p>Data output to both the console and text files can be supressed with Verbose=F and Output=F respectively. This is particularly useful when the user prefers to send the results to an R object (Return=T, see below) rather than to text files for further analysis. Finally, when multiple analyses are run (i.e., HWE, H, L, A) the data for each analysis (Chi Squares, Odds Ratios, Allele Frequencies, Allele Counts, etc.) can be written as set of files labeled ‘Merge’ for convenient access (Merge.Ouput=T).</p>
<p><em>Output As List Object</em><br />
With the pararmeter “Return=T”, the result is returned as a list with indices for each analysis result (HWE, H, L, A). Defined loci sets will exists as list sub-elements under each respective analysis. If no Loci.Sets were specified, only one list element will exist for Set1.</p>
<p>For an example. If BIGDAWG outputs to an object called ‘BD’ in R (see examples below).</p>
<p>BD (e.g., 3 loci sets defined)<br />
..$HWE - Results of the the Hardy-Weinberg equilibrium test<br />
..$H$Set1 - Results of the ‘Haplotype’ analysis for Set1 ..$H$Set2 - Results of the ‘Haplotype’ analysis for Set2<br />
..$H$Set3 - Results of the ‘Haplotype’ analysis for Set3<br />
..$L... - Results of the ‘Locus’ analysis<br />
..$A... - Results of the ‘Amino Acid’ analysis</p>
<p>or when “All.Pairwise=T” (e.g., 3 loci, 3 pairwise comparisons possible, single locus set)<br />
..$HWE - Results of the the Hardy-Weinberg equilibrium test<br />
..$H$Set1.PairwiseSet1 - Results of the ‘Haplotype’ analysis for Set1 Pairwise Set 1<br />
..$H$Set1.PairwiseSet2 - Results of the ‘Haplotype’ analysis for Set1 Pairwise Set 2<br />
..$H$Set1.PairwiseSet3 - Results of the ‘Haplotype’ analysis for Set1 Pairwise Set 3<br />
..$L$Set1 - Results of the ‘Locus’ analysis for Set1<br />
..$A$Set1 - Results of the ‘Amino Acid’ analysis for Set1</p>
<p>names(BD) - Display complete list of available sets to index. For a complete list of available pairwise results, use names(BD$H). This results list is written to the results directory as an R object called ‘Analysis.RData’ for later use.</p>
</div>
<div id="error-messages-and-codes" class="section level2">
<h2>Error Messages and Codes</h2>
<p>BIGDAWG has a few built-in checks to ensure data format consistency and compatibility, especially for HLA data. BIGDAWG also does a parameter review before performing chi-squared tests and returns ‘NCalc’ (not calculated) when all genotypes have expected counts &lt; 5 or the degrees of freedom do not allow for a test (e.g., dof &lt; 1).</p>
</div>
<div id="parameters" class="section level2">
<h2>Parameters</h2>
<p><code>BIGDAWG(Data, HLA=T, Run.Tests, Loci.Set, All.Pairwise=F, Trim=F, Res=2, Missing=0, Cores.Lim=1L, Results.Dir, Return=F, Output=T, Verbose=T)</code></p>
<p><strong>Data</strong></p>
<p>Class: String. Required. No Default.</p>
<p>e.g., <code>Data=&quot;HLA_data&quot;</code> -or- <code>Data=&quot;foo.txt&quot;</code>-or- <code>Data=foo</code></p>
<p>Specifies genotype data file name. May use file name within working directory or full file name path to specify file location. See Data Input section for details about file formatting. Use ‘Data=HLA_data’ to analyze the bundled synthetic dataset.</p>
<p><strong>HLA</strong></p>
<p>Class: logical. Optional. Default = T.</p>
<p>Indicates whether or not your data is specific for HLA loci. If your data is not HLA, is a mix of HLA and data for other loci, or includes non-standard HLA allele names, you should set <code>HLA = F</code>. This will override the Trim and EVS.rm arguments, and will skip various tests and checks. Set <code>HLA = T</code> if and only if the dataset HLA alleles name are consistent with the most recent <a href="https://www.ebi.ac.uk/ipd/imgt/hla/">IPD-IMGT/HLA</a> database release.</p>
<p><strong>Run.Tests</strong></p>
<p>Class: String or Character vector. Optional. Default = Run all tests.</p>
<p>e.g., <code>Run.Tests = c(&quot;L&quot;,&quot;A&quot;)</code> -or- <code>Run.Tests = &quot;HWE&quot;</code></p>
<p>Specifies which tests to run in analysis. “HWE” will run the Hardy Weinberg Equilibrium test, “H” will run the haplotype association test, “L” will run the locus association test, and “A” will run the amino acid association test. Combinations of the test are permitted as indicated in the example.</p>
<p>The amino acid test generally requires the most processing time. Taking advantage of multi-core machines can minimize this time (see below). Moreover, avoid defining multiple Loci.Sets (see below) with overlapping loci as processing time will increasing redundantly. Currently, the amino acid analysis is limited to the HLA-A, -B, -C, -DRB1, -DRB3, -DRB4, -DRB5, -DQA1, -DQB1, -DPA1 and -DPB1 loci.</p>
<p><strong>Loci.Set</strong></p>
<p>Class: List. Optional. Default = Use all loci.</p>
<p>e.g., <code>Loci.Set=list(c(&quot;DRB1&quot;,&quot;DQB1&quot;),c(&quot;A&quot;,&quot;DRB1&quot;,&quot;DPB1&quot;), c(&quot;DRB1&quot;,&quot;DRB3&quot;))</code></p>
<p>Input list defining which loci to use for analyses. Combinations are permitted. If you included HLA-DRB3,-DRB4,-DRB5 as a collapsed column pair (‘DRB3/4/5’), you must specifiy the single locus in the Loci.set if you wish them to be included in an analysis set (i.e., ‘DRB3’ <strong>NOT</strong> ‘DRB3/4/5’). The pair of alleles for a locus must be in adjacent columns in the analyzed data set.</p>
<p>Running multiple sets is generally only relevant for the haplotype analysis without all pairwise combinations. For all other analyses, loci are treated independently. Consider running haplotype analysis independently when optioning multi-locus sets that included overlapping loci to avoid redundancies. Each locus set’s output will be contained within a separate set folder numbered numerically corresponding to their order in the Loci.set parameter (see Data Output section).</p>
<p><strong>All.Pairwise</strong></p>
<p>Class: Logical. Optional. Default = F.</p>
<p>Should pairwise combinations of loci be run in the haplotype analysis? Only relevant to haplotype analysis. When optioned, only pairwise combinations of loci will be run and not all the loci in a given data set.</p>
<p><strong>EVS.rm</strong></p>
<p>Class: Logical. Optional. Default = F. (<code>HLA=T</code> specific).</p>
<p>Flags whether or not to strip expression variant suffixes from HLA alleles. Example: A*01:11N will convert to A*01:11. Should not be optioned for data that does not conform to HLA naming conventions.</p>
<p><strong>Trim</strong></p>
<p>Class: Logical. Optional. Default = F. (<code>HLA=T</code> specific).</p>
<p>Flags whether or not to Trim HLA alleles to a specified resolution. Should not be optioned for data that does not conform to HLA naming conventions.</p>
<p><strong>Resolution</strong></p>
<p>Class: Numeric. Optional. Default = 2. (<code>HLA=T</code> specific).</p>
<p>Sets the desired resolution when trimming HLA alleles. Used only when <code>Trim = T</code>. Fields for HLA formatting must follow current colon-delimited nomenclature conventions. Currently, the amino acid analysis will automatically truncate to 2-field resolution. Trimming is automatic and need not be optioned for amino acid analysis to run. This test will not run for data that does not conform to HLA naming conventions.</p>
<p><strong>Missing</strong></p>
<p>Class: String/Numeric. Optional. Default = 0.</p>
<p>Sets the allowable per subject threshold for missing alleles. Relevant to running the haplotype analysis. Effects can be disastrous on processing time and memory allocation for large values (&gt;2) of missing. Missing may be set as a number or as “ignore” to skip removal and retain all subjects. If you find BIGDAWG spending a lot time on the “Estimating Haplotypes…” step, reduced you missing to a value less than or equal to 2.</p>
<p><strong>Cores.Lim</strong></p>
<p>Class: Integer. Optional. Default = 1 Core.</p>
<p>Specifies the number of cores accessible by BIGDAWG in amino acid analysis. Not relevant to Windows operating systems which will use only a single core. More than 1 core is best when optioned in command line R and is not recommend when used in combination with a GUI, e.g. RStudio.</p>
<p><strong>Results.Dir</strong></p>
<p>Class: String. Optional. Default = see Data Output section.</p>
<p>String name of a folder for BIGDAWG output. Subfolder for each locus set will be generated within any output folder specified.</p>
<p><strong>Return</strong></p>
<p>Class: Logical. Optional. Default = F.</p>
<p>Specifies if BIGDAWG should output analysis results to a specified object.</p>
<p><strong>Output</strong></p>
<p>Class: Logical. Optional. Default = T.</p>
<p>Turns on or off the writing of results to files. The default will write all results to text files in the output directory.</p>
<p><strong>Merge.Output</strong></p>
<p>Class: Logical. Optional. Default = F.</p>
<p>Turns on or off the merging of all analysis results into single files labeled ’Merged_xxxx.txt&quot;. This process is rapid for smaller sets of loci (less than 50). However it can become increasingly cpu time intensive when there are many loci analyzed in conjuction with all pairwise combinations in the haplotype analysis. This parameter is subordinate to Output=T.</p>
<p><strong>Verbose</strong></p>
<p>Class: Logical. Optional. Default = T.</p>
<p>Sets the levels of detail that should be displayed on the console. The default will display summaries of the analysis from each specified test. When turned off, only the completion status of each test is displayed.</p>
</div>
<div id="examples" class="section level2">
<h2>Examples</h2>
<p>These are examples only and need not be run as defined below.</p>
<pre><code>
# Install the BIGDAWG package
install.packages(&quot;BIGDAWG&quot;)

# Run the full analysis using the example set bundled with BIGDAWG
BIGDAWG(Data=&quot;HLA_data&quot;)

# Run the haplotype analysis with all pairwise combinations on a file called &#39;data.txt&#39;
BIGDAWG(Data=&quot;data.txt&quot;, Run.Tests=&quot;H&quot;, All.Pairwise=T)

# Run the Hardy-Weinberg and Locus analysis with non-HLA data while ignoring any missing data on a file called &#39;data.txt&#39;
BIGDAWG(Data=&quot;data.txt&quot;, HLA=F, Run.Tests=c(&quot;HWE&quot;,&quot;L&quot;), Missing=&quot;ignore&quot;)

# Run the amino acid analysis trimming data to 2-Field resolution on a file called &#39;data.txt&#39;
BIGDAWG(Data=&quot;data.txt&quot;, Run.Tests=&quot;A&quot;, Trim=T, Res=2)

# Run the haplotype analysis with subsets of loci on a file called &#39;data.txt&#39;
BIGDAWG(Data=&quot;data.txt&quot;, Run.Tests=&quot;H&quot;, Loci.Set=list(c(&quot;DRB1&quot;,&quot;DQB1&quot;,&quot;DPB1&quot;),c(&quot;DRB1&quot;,&quot;DQB1&quot;)))

# Run the full analysis, minimize console output, disable write to file, output to object &#39;BD&#39;
BD &lt;- BIGDAWG(Data=&quot;data.txt&quot;, Output=F, Return=T, Verbose=F)
</code></pre>
</div>
<div id="updating-the-bundled-imgthla-protein-alignment" class="section level2">
<h2>Updating the bundled IMGT/HLA protein alignment</h2>
<p>For the amino acid analysis, BIGDAWG is bundled with HLA protein alignment data using theabove indicated database release. These bundled alignments can be updated to the most recent release <a href="https://www.ebi.ac.uk/ipd/imgt/hla/">IPD-IMGT/HLA</a>. Future database updates do not guarantee compatibility with the updating tool.</p>
<pre><code># Identify the installed and current release of the bundled IMGT/HLA database release
# Requires active internet connection.
CheckRelease()
CheckRelease(Package=F) # restricts to IMGT/HLA database versions only

# Update to the most recent IMGT/HLA database release
UpdateRelease()

# Force update
UpdateRelease(Force=T)

# Restore to original bundled update.
UpdateRelease(Restore=T)</code></pre>
</div>
<div id="updating-bigdawg-to-latest-developmental-versions" class="section level2">
<h2>Updating BIGDAWG to latest developmental versions</h2>
<p>Developmental versions of BIGDAWG can be downloaded through <a href="https://github.com/IgDAWG/BIGDAWG/">GitHub</a> or using the following code (requires R package ‘devtools’). These versions will include the most up-to-date bug fixes as well as access to new features that are still under development. GitHub versions are constantly under development and if you prefer a more stable fixed release, install BIGDAWG from the CRAN respository. You may check BIGDAWG versions using CheckRelease(). Local builing of vignettes requires pandoc and pandoc-citeproc if you do not use RStudio <a href="https://pandoc.org/installing.html">Pandoc</a>. Before installation from GitHub, it is recommended that all other R packages be up to date. May require a R/Rstudio close and restart.</p>
<pre><code># Identify the installed (local), release (CRAN), and developmental (GitHub) versions of BIGDAWG.
# Requires active internet connection.
CheckRelease()
CheckRelease(Alignment=F) # restricts to BIGDAWG versions only

# If &#39;devtools&#39; package installation is required for installation via GitHub.
install.packages(&quot;devtools&quot;)

# Load latest BIGDAWG version from GitHub
# May require closing and reopening of R Studio after install
library(&quot;devtools&quot;)
install_github(&quot;IgDAWG/BIGDAWG&quot;, build_vignettes = TRUE) # Requires Pandoc or RStudio
install_github(&quot;IgDAWG/BIGDAWG&quot;) # No Pandoc or RStudio

# For a temporary install of the developmental version
library(&quot;devtools&quot;)
dev_mode(on=T)
install_github(&quot;IgDAWG/BIGDAWG&quot;)
# .... run BIGDAWG analysis
dev_mode(on=F)</code></pre>
<p><em>End of vignette.</em></p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
